storage/src/main/scala/coop/rchain/storage/Storage.scala

```diff
@@ -1,266 +1,76 @@
-
-/*
-Storage is a key-value database that provides a Prolog-like unification
-mechanism for querying.  More specifically, it stores data in an Lmdb object.
-
-Storage accepts keys and values of type:
- * primitive Java data types (and thus no unsigned types)
- * Java String
- * Key (defined in this package)
-
-Only keys of type Key are eligible for unification.
-
-In this implementation, all keys that can be unified must be of type Key.
-
-Just like the Lmdb class, the Storage class can be configured:
- * read-only or writable
- * a key is associated with a single key or multiple keys
-
- */
-
 package coop.rchain.storage
 
-import coop.rchain.storage.Bb.Bbable
-import java.io.File
 import java.nio.ByteBuffer
-
-import coop.rchain.storage.QueryTools.{createKeySubstition}
-import org.lmdbjava.Txn
-
-import scala.collection.mutable.{ArrayBuffer, LinkedHashSet}
-import scala.io.Source
-
-class Storage(storConf: StorageConfig) {
-
-  if (!storConf.isValid())
-    throw new RChainException("Storage: StrorageConfig is invalid")
-
-  protected[storage] val lmdb =
-    new Lmdb(storConf.dirName,
-             storConf.name,
-             storConf.isKeyToValues,
-             storConf.isWritable,
-             storConf.baseDir,
-             storConf.maxSize)
-
-  // uniKeys returns a list of all keys (of type Key) in storage.
-  def uniKeys: java.util.Iterator[Key] = {
-    termTreeKeysSorted.iterator
-  }
-  protected[storage] val termTreeKeysSorted = new java.util.TreeSet[Key]
-
-  // returns an array of keys that unify with the query
-  def unifyQuery(query: Key): Array[Key] = {
-    var bindings = LinkedHashSet[Array[Binding]]()
-
-    val keyIter = uniKeys
-    while (keyIter.hasNext()) {
-      val key = keyIter.next()
-      val keyName = key.name
-      val keyParams = key.params
-      if (query.name == keyName && query.arity == key.arity) {
-        val (success, bindingsToAdd) =
-          QueryTools.unifyParams(query.params.params, keyParams.params)
-        if (success) {
-          bindings += bindingsToAdd
-        }
-      }
-    }
-
-    var keys = new ArrayBuffer[Key]()
-    for (bindingsArray <- bindings) {
-      val keySub = createKeySubstition(query, bindingsArray)
-      val valuesOption = getStrings(keySub.term)
-      if (valuesOption.isDefined) {
-        keys += keySub
+import java.nio.file.Path
+
+import cats.syntax.either._
+import coop.rchain.storage.util._
+import org.lmdbjava.DbiFlags.MDB_CREATE
+import org.lmdbjava.{Dbi, Env, Txn}
+
+class Storage private (env: Env[ByteBuffer], db: Dbi[ByteBuffer])
+    extends IStorage
+    with AutoCloseable {
+
+  def put[A](key: Key, value: A)(implicit s: Serialize[A]): Either[Error, Unit] =
+    Either
+      .catchNonFatal {
+        val encoded = s.encode(value)
+        val keyBuff = ByteBuffer.allocateDirect(env.getMaxKeySize)
+        val valBuff = ByteBuffer.allocateDirect(encoded.length)
+        ignore { keyBuff.put(key.bytes).flip() }
```

Comment:
Why use `ignore`?

Reply:
I want to be explicit about the fact that this (impure) expression returns a value (in this case for the sake of method-chaining) that I am discarding, and am instead only using for its side-effects.

We might imagine doing something like this:
```scala
val _ = keyBuff.put(key.bytes).flip()
val _ = valBuff.put(encoded).flip()
```

However, `_` is a normal identifier in Scala, and we can't rebind variable names.  So instead we have to do something like:
```scala
val _1 = keyBuff.put(key.bytes).flip()
val _2 = valBuff.put(encoded).flip()
```
...which is kind of awful, and will also trigger compiler and IDE warnings about unused variables. 

So instead, I borrowed the `ignore` function from the OCaml standard library.

I'd also like to start using the [wartremover](http://www.wartremover.org/) code linting tool on our code, which will complain about cases where we are discarding return values without explicitly accounting for doing so.